File: occasum-core/src/main/java/pw/stamina/occasum/properties/ParameterizedProperty.java
package pw.stamina.occasum.properties;

import java.util.Objects;

public abstract class ParameterizedProperty<T> extends AbstractProperty {
    private T value;
    protected final T defaultValue;

    protected ParameterizedProperty(String name, T value) {
        super(name);

        Objects.requireNonNull(value, "value");
        this.value = value;
        this.defaultValue = value;
    }

    public final T get() {
        return value;
    }

    public void set(T value) {
        Objects.requireNonNull(value, "value");
        this.value = value;
    }

    @Override
    public final boolean isDefault() {
        return value.equals(defaultValue);
    }

    public final T getDefault() {
        return defaultValue;
    }

    @Override
    public final void reset() {
        set(defaultValue);
    }

    @Override
    public final String getValueAsString() {
        return valueToString(value);
    }

    @Override
    public final String getDefaultValueAsString() {
        return valueToString(defaultValue);
    }

    protected String valueToString(T value) {
        return String.valueOf(value);
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/properties/ParameterizedProperty.kt
package pw.stamina.occasum.properties

abstract class ParameterizedProperty<T> protected constructor(name: String, private var value: T) : AbstractProperty(name) {
    val default: T = value

    override val isDefault: Boolean
        get() = value == default

    override val valueAsString: String
        get() = valueToString(value)

    override val defaultValueAsString: String
        get() = valueToString(default)

    fun get(): T? {
        return value
    }

    open fun set(value: T) {
        this.value = value
    }

    override fun reset() {
        set(default)
    }

    protected open fun valueToString(value: T): String {
        return value.toString()
    }
}
